### PR TITLE
MAINT: Fix computation of numpy.array_api.linalg.vector_norm

### LINKED ISSUE #21083: BUG: "Non-standard" vector-norms of 1-D arrays in norm() always return float64 dtype
### Describe the issue:

The "non-standard" norms in `norm()`, i.e., `ord < 0` or `ord > 2` on a 1-d array (so it's a vector norm and not a matrix norm), incorrectly promote the result to float64 regardless of the input dtype. This is because the code just manually implements it with

```py
absx = abs(x)
absx **= ord
ret = add.reduce(absx, axis=axis, keepdims=keepdims)
ret **= (1 / ord)
return ret
```

at https://github.com/numpy/numpy/blob/1168868df63678e5704acd866fafcf40dc849481/numpy/linalg/linalg.py#L2559.

Specifically, the last line `ret **= (1 / ord)` promotes the result to float64 because `ret` is a scalar and `1/ord` is a Python float. 

### Reproduce the code example:

```python
>>> np.linalg.norm(np.asarray([1., 2.], dtype=np.float32), ord=-1).dtype
dtype('float64')
>>> np.linalg.norm(np.asarray([1., 2.], dtype=np.float16), ord=-1).dtype
dtype('float64')
>>> np.linalg.norm(np.asarray([1., 2.], dtype=np.float32), ord=1).dtype
dtype('float32')
>>> np.linalg.norm(np.asarray([1., 2.], dtype=np.float16), ord=1).dtype
dtype('float16')
```

### Error message:

_No response_

### NumPy/Python version information:

>>> np.__version__
'1.23.0.dev0+782.g1168868df6'

### PR DESCRIPTION
**Pull request by asmeurer**
There were several issues due to the fact that it was mostly untested in the array API test suite. 

This also fixes https://github.com/numpy/numpy/issues/21083. I'm unclear exactly how I should add a test for this. 

For the array API vector_norm, tests are being added to the array API test suite.

### CODE REVIEW

File: numpy/array_api/linalg.py [hunk 1]
```diff
@@ -391,18 +392,43 @@ def vector_norm(x: Array, /, *, axis: Optional[Union[int, Tuple[int, int]]] = No
...
+    if keepdims:
+        # We can't reuse np.linalg.norm(keepdims) because of the hacks above
+        # to avoid matrix norm logic.
+        shape = list(x.shape)
+        if axis is None:
+            _axis = range(x.ndim)
+        elif isinstance(axis, int):
+            _axis = (axis,)
+        else:
+            _axis = axis
```
Comment (seberg) [inline numpy/array_api/linalg.py:427]: Might put this together with the setup (i.e. figuring out the result shape there?  But looks fine.

**Review by seberg — COMMENTED** [review]
A few small comments/suggestions but looks fine.  Maybe @honno can sign this off?

**Comment by asmeurer** [comment]
@seberg any thoughts on how I can add a test for #21083?

**Comment by seberg** [comment]
Frankly, no, this is almost certainly ancient code from the early days of NumPy (or possibly older!), I would just git grep for `norm` tests and add one next to it where it seems reasonable.

**Comment by asmeurer** [comment]
Yeah, I figured out where the norm tests are. I just am really unclear how exactly they work. Just reading them it looks to me like this ought to already be tested (specifically [here](https://github.com/numpy/numpy/blob/c92739b05e4b6aee621b8216ff45cc47a62210c7/numpy/linalg/tests/test_linalg.py#L1553)).

Actually, looking closer now, I *think* I see the problem, which is that the test [here](https://github.com/numpy/numpy/blob/c92739b05e4b6aee621b8216ff45cc47a62210c7/numpy/linalg/tests/test_linalg.py#L1244) doesn't actually test the output dtype, just that it is a floating dtype.

**Comment by seberg** [comment]
On first glance, I think the tests are just buggy. `_TestNormGeneral` doesn't seem to actually use `self.dt` in the relevent tests?  And even then, I am not sure its checks would be sufficient (but didn't dig in).

It would be nice to just replace this stuff with `pytest.parametrize`, that would be *much* clearer to read than this inheritance dance.

**Comment by asmeurer** [comment]
I agree. I don't know if this sort of thing is standard in the NumPy test suite or if it's just a weird leftover thing from ages ago that hasn't been modernized.

**Comment by seberg** [comment]
Just a weird leftover, probably it was a good and typical pattern at some point.  I would not mind if it was just parametrized even if that is a biggish diff, on first sight, it looks straight forward enough.

File: numpy/array_api/linalg.py [hunk 2]
```diff
@@ -391,18 +392,43 @@ def vector_norm(x: Array, /, *, axis: Optional[Union[int, Tuple[int, int]]] = No
...
+    # np.linalg.norm tries to do a matrix norm
     a = x._array
     if axis is None:
-        a = a.flatten()
-        axis = 0
+        # Note: np.linalg.norm() doesn't handle dimension 0 arrays
+        if a.ndim == 0:
+            a = a[None]
+        else:
+            a = a.flatten()
```
[the remark this answers is not in the document]
Comment (seberg) [reply numpy/array_api/linalg.py:402]: Yes, but `np.array(0).ravel()` (or flatten) also is 1D, this is not `<= 1` but `== 1`, so `ravel()` covers the `a[None]` path just as well.

**Comment by asmeurer** [comment]
I also didn't realize those test are also testing integer inputs to norm() (which aren't even in scope in the array API). So testing the correct dtype is a little more complicated. Is there a helper function to get the correct promoted type? 

Also, I hope this isn't a problem, but I'm probably not going to refactor those tests to use parameterization. It would make a good first issue for someone to do that, but I really just want to add a test for my fix if it's not too complicated.

**Comment by seberg** [comment]
> Is there a helper function to get the correct promoted type?

I dunno `np.result_type(dtype1, dtyp2, 0.)`?  I am not sure what the current logic ends up doing exactly, though.  I mean, promotion does depend on the operation in principle, i.e. the maximum norm should preserve dtype in principle, but other norms should promote to floating.

> Also, I hope this isn't a problem, but I'm probably not going to refactor those tests to use parameterization. It would make a good first issue for someone to do that, but I really just want to add a test for my fix if it's not too complicated.

nah sure, but doesn't hurt to hope ;).

**Comment by honno** [comment]
@seberg's comments on using `ravel()` and the existing axes utils (if possible) make sense, otherwise LGTM. Personally I prefer figuring out result shapes after the fact, as you can follow up the reshape straight after to reduce cognitive load.

**Comment by asmeurer** [comment]
I think fiddling with the existing linalg tests is going to be a deeper rabbit hole than I really want to get down for this PR. If you want I can add a simple, separate regression test for the norm() fix. Otherwise, I would suggest opening a "good first contribution" issue for refactoring those tests to use pytest parameterization, and also to more explicitly test the result type in each case.

I have cleaned up the implementation of numpy.array_api.vector_norm a little bit as per the review comments.

**Comment by seberg** [comment]
Unfortunately there is now a (probably very small) merge conflict.  @honno otherwise you were happy with this and we should just merge it?

**Comment by honno** [comment]
> Unfortunately there is now a (probably very small) merge conflict. @honno otherwise you were happy with this and we should just merge it?

Yep LGTM. Again @seberg's previous suggestions on `ravel()` and exploring existing axis are worth exploring, but it'd be nice to just get this in.

However let's just see if @asmeurer is okay with this PR still, as there have been more questions and clarifications on the `linalg` extension in the spec since this PR was opened.

**Comment by asmeurer** [comment]
Looks like https://github.com/numpy/numpy/issues/21083 was fixed by  https://github.com/numpy/numpy/pull/17709 so my fix is no longer needed. No opinion on which of the fixes is better. I guess neither would be needed once type promotion is cleaned up. At least that PR was (apparently) able to add tests.

**Comment by seberg** [comment]
Lets put it in then.  Thanks!

**Pull request merged**